api/src/main/java/org/openmrs/api/StorageService.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.stream.Stream;
+
+/**
+ * Storage service to persist data that does not fit well in DB.
+ * <p>
+ * Data objects are stored under generated or provided unique keys. Data and keys are immutable.
+ * 
+ * @since 2.8.0, 2.7.2, 2.6.16, 2.5.15
```

Comment:
Ideally we backport the interface and implementation to all versions used by o3, o2, bahmni, ozone... It is to be able to easily add support in modules. We will only modify the master branch to use the new service in api. Does this list of versions cover what is used by implementations these days?

Reply:
The thing is that if you want to correct your module to use storage service and it's only available in 2.8.0 then your module is incompatible with all older versions. It's not something module authors wanted to do historically. If we backport it to older branches then modules only need to require new maintenance branches of core, but do not have to suddenly drop support for core older than 2.8.0.

Reply:
> It's not something module authors wanted to do historically

If that's the problem we're trying to solve, we'll need to backport to a lot further. While I don't know of any major deployments running against 2.3.x or lower, for example, very few modules are written that target anything newer than that and many of them target substantially older versions. E.g., attachments targets 2.2.0, module-spa targets 2.4.1, reporting targets 1.9.9.

I think the best path here is to add this to 2.8.0 and let module authors figure out how they want to handle the migration, e.g., force the module to only be compatible with 2.8.0 or add their own abstraction layer that handles both.

Reply:
It's tricky to add the abstraction layer... I'm still hoping to have it back ported to make the migration easier. Let's discuss on platform call.

Reply:
I actually like experimenting with this new approach as an alternative to the mess we have been going through to support multiple platform versions. How far to backport would not be a big deal for now because if at any one time we find that we need a lower back port, all we need is do it and release its corresponding maintenance version. As simple as that. 😊

Reply:
All right. Could I backport it at least as far as 2.5.x for now so implementations on refapp 2.x can use it?

Reply:
No objection from me. 👍

Reply:
`2.7.2` has moved to `2.7.4`